### PR TITLE
Add checkmode support

### PR DESCRIPTION
**Pull request by SimonBarendse**
When running your playbooks in [checkmode](https://docs.ansible.com/ansible/latest/user_guide/playbooks_checkmode.html), the modules now return whether they will make changes.

### CODE REVIEW

File: library/secrethub_read.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def __init__(self):
                 'type': "str",
             },
         }
-        super(ReadModule, self).__init__(argument_spec)
+        super(ReadModule, self).__init__(argument_spec, supports_check_mode=True)
 
     def run(self):
         try:
```
Comment (mackenbach) [inline library/secrethub_read.py:74]: Should it perform a `read` when running in check mode (i.e. triggering an audit event and decrypting data)? In the other modules, your design looks like it only performs client calls when not in check mode.

Comment (mackenbach) [reply library/secrethub_read.py:74]: Also, what happens when the requested secret does not exist? Should it fail in check mode then?

File: library/secrethub_generate.py [hunk 2]
```diff
@@ -95,29 +95,28 @@ def __init__(self):
...
-            self.fail_json(
-                changed=False,
-                msg=str(e),
-            )
-        except ReadError as e:
-            self.fail_json(
-                changed=True,
-                msg=str(e),
-            )
+        if not self.check_mode:
```
Comment (mackenbach) [inline library/secrethub_generate.py:101]: What happens when the parent directory (or repo, access, etc.) does not exist and you're running this in check mode? Is the expected behavior that it fails or that it reports everything works fine?

**Review by mackenbach — COMMENTED** [review]
Everything looks good to me, very nice addition to the module. I have a few questions on the expected behavior of the `check mode`, have a look.

File: library/secrethub_read.py (hunk 1 above)
Comment (SimonBarendse) [reply library/secrethub_read.py:74]: > Also, what happens when the requested secret does not exist? Should it fail in check mode then?

For example the `copy` module, fails in check mode when the source file does not exists: https://github.com/ansible/ansible/blob/v2.7.4/lib/ansible/modules/files/copy.py#L320

> Should it perform a read when running in check mode (i.e. triggering an audit event and decrypting data?

I think triggering audit events is a side-effect we should allow the module to have. This holds for running in check-mode, but they are also an exception to the idempotent behavior of the module: triggering the module multiple times will create multiple audit events. All other state will be untouched in subsequent runs.

However, in check-mode, it is not necessary to fetch the actual secret data. Checking that the secret version exists is enough. We could do so by doing an ls on the secret path (including version if given).
By doing so, the module does not do more than required and the audit events that are created when running in check-mode align better with the actions performed.

File: library/secrethub_generate.py (hunk 2 above)
Comment (SimonBarendse) [reply library/secrethub_generate.py:101]: Good point! Yes, the module should fail in that case.

We can use the same solution as I just proposed for the read module: we can do an `ls` on the parent path to check for existence.

File: library/secrethub_read.py (hunk 1 above)
Comment (mackenbach) [reply library/secrethub_read.py:74]: Alright cool! So if I get this correctly the module will do the following in `check` mode:
- Fail if a required condition is not met
- Only check if a secret exists without reading it

Sounds like pretty predictable behavior, I like it 👍

**Comment by mackenbach** [comment]
Closing this PR as it is stale. @SimonBarendse it can be reopened when we pick this up again.

**Pull request closed without merging**